
- fail gracefully if a 1.3 config file is found
- use links not lynx, and fail gracefully when neither is found
- source sysconfig/httpd for custom env. vars etc.
- make httpd -t work even in SELinux
- refuse to restart into a bad config (done better on HEAD)

Upstream-HEAD: vendor
Upstream-2.0: vendor
Upstream-Status: Vendor-specific changes for better initscript integration

--- httpd-2.0.52/support/apachectl.in.apctl
+++ httpd-2.0.52/support/apachectl.in
@@ -41,32 +41,27 @@
 # 
 # the path to your httpd binary, including options if necessary
 HTTPD='@exp_sbindir@/@progname@'
-#
-# pick up any necessary environment variables
-if test -f @exp_sbindir@/envvars; then
-  . @exp_sbindir@/envvars
-fi
+
 #
 # a command that outputs a formatted text version of the HTML at the
 # url given on the command line.  Designed for lynx, however other
 # programs may work.  
-LYNX="lynx -dump"
+if [ -x /usr/bin/links ]; then
+   LYNX="links -dump"
+elif [ -x /usr/bin/lynx ]; then
+   LYNX="lynx -dump"
+else
+   LYNX="none"
+fi
+
 #
 # the URL to your server's mod_status status page.  If you do not
 # have one, then status and fullstatus will not work.
 STATUSURL="http://localhost:@PORT@/server-status"
-#
-# Set this variable to a command that increases the maximum
-# number of file descriptors allowed per child process. This is
-# critical for configurations that use many file descriptors,
-# such as mass vhosting, or a multithreaded server.
-ULIMIT_MAX_FILES="@APACHECTL_ULIMIT@"
-# --------------------                              --------------------
-# ||||||||||||||||||||   END CONFIGURATION SECTION  ||||||||||||||||||||
-
-# Set the maximum number of file descriptors allowed per child process.
-if [ "x$ULIMIT_MAX_FILES" != "x" ] ; then
-    $ULIMIT_MAX_FILES
+
+# Source /etc/sysconfig/httpd for $HTTPD setting, etc.
+if [ -r /etc/sysconfig/httpd ]; then
+   . /etc/sysconfig/httpd
 fi
 
 ERROR=0
@@ -74,23 +69,65 @@
     ARGV="-h"
 fi
 
+function check13() {
+# check for 1.3 configuration
+GONE="(ServerType|BindAddress|Port|AddModule|ClearModuleList|"
+GONE="${GONE}AgentLog|RefererLog|RefererIgnore|FancyIndexing|"
+GONE="${GONE}AccessConfig|ResourceConfig)"
+if grep -Eiq "^[[:space:]]*($GONE)" /etc/httpd/conf/httpd.conf; then
+  echo "$0: Apache 1.3 configuration directives found"
+  echo "$0: please read @docdir@/migration.html"
+  exit 2
+fi
+}
+
+function checklynx() {
+if [ "$LYNX" = "none" ]; then
+   echo "The 'links' package is required for this functionality."
+   exit 8
+fi
+}
+
+function testconfig() {
+# httpd is denied terminal access in SELinux, so run in the
+# current context to get stdout from $HTTPD -t.
+if test -x /usr/sbin/selinuxenabled && /usr/sbin/selinuxenabled; then
+  runcon -- `id -Z` $HTTPD -t
+else
+  $HTTPD -t
+fi
+ERROR=$?
+}
+
 case $ARGV in
-start|stop|restart|graceful)
+restart|graceful)
+    if $HTTPD -t >&/dev/null; then
+       $HTTPD -k $ARGV
+       ERROR=$?
+    else
+       echo "apachectl: Configuration syntax error, will not run \"$ARGV\":"
+       testconfig
+    fi
+    ;;
+start|stop)
+    check13
     $HTTPD -k $ARGV
     ERROR=$?
     ;;
 startssl|sslstart|start-SSL)
+    check13
     $HTTPD -k start -DSSL
     ERROR=$?
     ;;
 configtest)
-    $HTTPD -t
-    ERROR=$?
+    testconfig
     ;;
 status)
+    checklynx
     $LYNX $STATUSURL | awk ' /process$/ { print; exit } { print } '
     ;;
 fullstatus)
+    checklynx
     $LYNX $STATUSURL
     ;;
 *)
